File: app/src/main/java/com/example/realweather/view/DashboardFragment.java
package com.example.realweather.view;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.fragment.NavHostFragment;

import com.example.realweather.R;
import com.example.realweather.databinding.DashboardFragmentBinding;
import com.example.realweather.model.DisplayValueConverter;
import com.example.realweather.repository.PreferencesClient;
import com.example.realweather.viewmodel.DashboardViewModel;

public class DashboardFragment extends Fragment implements PreferencesClient {

    private DashboardViewModel viewModel;


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        OnBackPressedCallback callback = new OnBackPressedCallback(true /* enabled by default */) {
            @Override
            public void handleOnBackPressed() {
                requireActivity().finish();
            }
        };
        requireActivity().getOnBackPressedDispatcher().addCallback(this, callback);

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        DashboardFragmentBinding binding = DataBindingUtil.inflate(inflater,
                R.layout.dashboard_fragment, container, false);
        ForecastAdapter forecastAdapter = new ForecastAdapter(getUnitUserPreference(requireContext()));
        binding.forecastRecyclerView.setAdapter(forecastAdapter);
        binding.forecastRecyclerView.setHasFixedSize(true);
        requireActivity().setTitle(R.string.appName);
        setupViewModel();
        viewModel.getForecast().observe(this, forecastAdapter::setList);
        viewModel.getTodayForecast().observe(this, forecast -> {
            if (forecast == null) return;
            binding.setModel(forecast);
            binding.setConverter(new DisplayValueConverter(getUnitUserPreference(requireContext())));
        });
        setHasOptionsMenu(true);
        return binding.getRoot();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.settings) {
            NavHostFragment.findNavController(this).navigate(R.id.settingsFragment);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setupViewModel() {
        viewModel = ViewModelProviders.of(this).get(DashboardViewModel.class);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_main, menu);
    }


}

File: app/src/main/java/com/example/realweather/view/DashboardFragment.kt
package com.example.realweather.view

import android.os.Bundle
import android.view.*
import androidx.activity.OnBackPressedCallback
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.navigation.fragment.NavHostFragment
import com.example.realweather.R
import com.example.realweather.databinding.DashboardFragmentBinding
import com.example.realweather.model.DisplayValueConverter
import com.example.realweather.model.Forecast
import com.example.realweather.model.TodayForecast
import com.example.realweather.repository.PreferencesClient
import com.example.realweather.viewmodel.DashboardViewModel
import com.google.android.gms.ads.AdRequest

class DashboardFragment : Fragment(), PreferencesClient {
    private val viewModel: DashboardViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val callback: OnBackPressedCallback =
            object : OnBackPressedCallback(true /* enabled by default */) {
                override fun handleOnBackPressed() {
                    requireActivity().finish()
                }
            }
        requireActivity().onBackPressedDispatcher.addCallback(this, callback)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        (requireActivity() as MenuHost).addMenuProvider(object : MenuProvider {
            override fun onPrepareMenu(menu: Menu) {
                // Handle for example visibility of menu items
            }

            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.menu_main, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                if (menuItem.itemId == R.id.settings) {
                    NavHostFragment.findNavController(this@DashboardFragment)
                        .navigate(R.id.settingsFragment)
                    return true
                }
                return false
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = DataBindingUtil.inflate<DashboardFragmentBinding>(
            inflater,
            R.layout.dashboard_fragment, container, false
        )
        val forecastAdapter = ForecastAdapter(getUnitUserPreference(requireContext()))
        binding.forecastRecyclerView.adapter = forecastAdapter
        binding.forecastRecyclerView.setHasFixedSize(true)
        viewModel.forecast.observe(viewLifecycleOwner) { newList: List<Forecast>? ->
            forecastAdapter.setList(
                newList
            )
        }
        viewModel.todayForecast.observe(viewLifecycleOwner) { forecast: TodayForecast? ->
            if (forecast == null) return@observe
            binding.model = forecast
            binding.converter = DisplayValueConverter(getUnitUserPreference(requireContext()))
        }
        initializeAds(binding)
        return binding.root
    }

    private fun initializeAds(binding: DashboardFragmentBinding) {
        val adRequest = AdRequest.Builder()
            .build()
        binding.adView.loadAd(adRequest)
    }
}
